Add bulk editing support by passing array down to the fields

louwie17 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

Allows consumers to pass in an array of items to be edited. The panel view will show "Mixed" when the values are different across the items.

## Why?

Partially addresses: https://github.com/WordPress/gutenberg/issues/65685
The header for bulk editing will be addressed in a separate PR.

## How?

This is an alternative solution to: https://github.com/WordPress/gutenberg/pull/67344

DataForm allows an array of items to be passed in.
When an array of items is passed in it does a couple things:
- The `data` prop now supports `Item` and `Item[]`
- The `render` method receives a `value` prop which can be the Mixed symbol.
- The data form layouts call `getValue` and pass the value down. They also set it to Mixed if bulk editing and the value isn't the same across all records.

### Some questions/follow up:

- The `render` method takes a single `item` prop, how do we handle this when passing in an array of items? Should we replace this with the `data` and `value` props similar to the `edit` method? Or add support for an `items` prop?

## Testing Instructions

1. Enable the "Quick Edit in DataViews" experiment
2. Enable the 2024 or 2025 theme and go to **Appearance > Editor > Pages**
3. Change the view in the top right to **Table** and open the right side bar
4. Select multiple pages, it should show the title, status, author, and date fields
5. The fields should show **Mixed** if the values are different across pages.
6. Update the status for example, it should remove **Mixed** and show the selected status.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/e19b2c43-c0fa-4675-951f-8c5d76d3ceba

louwie17 committed:
Initial bulk editing changes

louwie17 committed:
Update fields to allow for array of form items

louwie17 committed:
Add type to disable bulk editing

louwie17 commented on packages/dataviews/src/types.ts:
```diff
@@ -179,11 +186,15 @@ export type Fields< Item > = Field< Item >[];
-	data: Item;
+export type DataFormControlProps<
+	Item,
+	SupportsBulkEditing extends boolean = true,
```
I added this additional type to make it easier to handle within fields that don't support bulk editing, but maybe there is a better way?

gigitux replied:
It is just a personal preference, but I would prefer type composition:

```ts
export type DataFormControlProps< Item > = {
	data: Item;
	field: NormalizedField< Item >;
	onChange: ( value: Record< string, any > ) => void;
	hideLabelFromVision?: boolean;
	value: any;
};

export type WithBulkEditing< Item > = {
	data: Item | Item[];
};
```

And use in this way:

```ts
export const ParentEdit = ( {
	data,
	field,
	onChange,
}: DataFormControlProps< BasePost > & WithBulkEditing< BasePost > ) => {
```

We could do export already the composed type:

```ts
export type DataFormControlWithBulkEditingProps< Item > = DataFormControlProps< Item > & WithBulkEditing< Item >;

export const ParentEdit = ( {
	data,
	field,
	onChange,
}: DataFormControlWithBulkEditingProps< BasePost > 
```

louwie17 commented on packages/fields/src/fields/slug/slug-edit.tsx:
```diff
@@ -27,7 +27,7 @@ const SlugEdit = ( {
+}: DataFormControlProps< BasePost, false > ) => {
 	const { id } = field;
 
 	const slug = field.getValue( { item: data } ) || getSlug( data );
```
If slug where to support bulk edits how would we handle the `getSlug` call here.
`field.getValue( { item: data } )` can be replaced by the `value` prop.

gigitux commented on packages/dataviews/src/dataforms-layouts/regular/index.tsx:
```diff
@@ -59,6 +60,28 @@ export default function FormRegularField< Item >( {
+			const intersects = remainingRecords.every( ( item ) => {
+				return fieldDefinition.getValue( { item } ) === firstValue;
+			} );
+			return intersects ? firstValue : MIXED_VALUE;
```
In `regular` layout we are passing the Symbol as value prop. I'm not sure that this is handled in the right way in the existent fields. For instance the feature image field:

https://github.com/WordPress/gutenberg/blob/0228bbc20978e603612b4e7da2bd94edddbad523/packages/fields/src/fields/featured-image/featured-image-edit.tsx#L26-L32

gigitux commented on packages/edit-site/src/components/post-edit/index.js:
```diff
@@ -24,17 +24,36 @@ import { privateApis as blockEditorPrivateApis } from '@wordpress/block-editor';
+	type: 'panel',
+	fields: [
+		{
+			id: 'featured_media',
+			layout: 'regular',
```
In this branch, the `featured_media` field isn't rendered:

<img width="1217" alt="image" src="https://github.com/user-attachments/assets/3bf68177-772f-48d6-bead-3e51863e22c6">

gigitux commented on packages/fields/src/fields/featured-image/featured-image-edit.tsx:
```diff
@@ -17,14 +17,12 @@ import { __ } from '@wordpress/i18n';
 	field,
 	onChange,
+	value,
 }: DataFormControlProps< BasePost > ) => {
```
Featured Image doesn't support the bulk editing, should we set `DataFormControlProps< BasePost, false >`?

gigitux commented on packages/dataviews/src/normalize-fields.ts:
```diff
@@ -64,6 +64,12 @@ export function normalizeFields< Item >(
 
+		let supportsBulkEditing = true;
+		// If custom Edit component is passed in we default to false for bulk edit support.
+		if ( typeof field.Edit === 'function' || field.supportsBulkEditing ) {
```
I'm not sure about this logic.

I think that we need to decide whether bulk editing support should be opt-in or opt-out, regardless of whether the Edit function is present.

louwie17 replied:
> I think that we need to decide whether bulk editing support should be opt-in or opt-out, regardless of whether the Edit function is present.

Yeah I agree! I would lean towards `optOut`

gigitux reviewed (commented):
Thanks for working on this! Overall the approach LGTM! I left a few comments

louwie17 committed:
Add WithBulkEditing type

louwie17 committed:
Change bulk editing to opt in

youknowriad commented on packages/dataviews/src/dataform-controls/radio.tsx:
```diff
@@ -7,16 +7,15 @@ import { useCallback } from '@wordpress/element';
 	hideLabelFromVision,
-}: DataFormControlProps< Item > ) {
+	value,
+}: DataFormControlProps< Item > & WithBulkEditing< Item > ) {
```
I don't understand this change. maybe I'm missing something when reading the diff though. Is "data" optional? where does "value" come from? I mean where are these type changes defined?

louwie17 replied:
I updated the `types.ts` with the above: https://github.com/WordPress/gutenberg/pull/67584/files#diff-05cd2939bbf237a5dbcd116a0da9c9019b05ec1ceaa9d429c5fc872f149f63c4R192-R197

`data` is still passed in, but is either `Item` or `Item | Item[]` depending if combined with `WithBulkEditing< Item >`

youknowriad replied:
Do we still need data?

louwie17 replied:
Yeah we still need `data` as there are still fields that rely on other data. Like the `parent` field which makes use of the post id and post type ( from the data object ) to filter the parent list. Or the slug field that uses `permalink_template` and `link` properties as well.

Currently I don't display these fields in bulk edit mode, as the logic may change a bit if there are multiple records passed down.
Unless there is a better way to pass these additional values down.

youknowriad commented on packages/dataviews/src/dataform-controls/datetime.tsx:
```diff
@@ -34,7 +33,7 @@ export default function DateTime< Item >( {
 			) }
 			<TimePicker
-				currentTime={ value }
+				currentTime={ typeof value === 'symbol' ? '' : value }
```
I don't see the check about symbol in other controls like integer, why?

louwie17 replied:
I will update it, the TimePicker component it-self seemed to handle it gracefully, but its not expected.

youknowriad commented on packages/dataviews/src/dataforms-layouts/panel/index.tsx:
```diff
@@ -98,6 +99,22 @@ function PanelDropdown< Item >( {
+		return fieldDefinition.getValue( {
+			item: data,
+		} );
+	}, [ data, fieldDefinition ] );
```
If I'm not wrong this merging is probably duplicated in "regular" too. Should we extract it to a utility?

louwie17 replied:
Correct, I can move it to a utility 👍

youknowriad commented on packages/dataviews/src/dataforms-layouts/panel/index.tsx:
```diff
@@ -138,7 +157,13 @@ function PanelDropdown< Item >( {
 				>
-					<fieldDefinition.render item={ data } />
+					{ showMixedValue ? (
+						__( 'Mixed' )
```
In one of the issues, @jasmussen mentions the possibility to render a different "mixed" label/component based on the "type" of the control. So potentially we could defer this to the control/field itself at some point.

jasmussen replied:
The context is loosely discussed [here](https://github.com/WordPress/gutenberg/issues/55101#issuecomment-2527826801).

youknowriad commented on packages/dataviews/src/normalize-fields.ts:
```diff
@@ -76,6 +81,7 @@ export function normalizeFields< Item >(
 			sort,
 			isValid,
 			Edit,
+			supportsBulkEditing,
```
Should we mark supportsBulkEditing as required in the "NormalizedField" type?

oandregal commented:
Sharing some early thoughts after having given this a quick test (haven't looked at the code yet):

- UI-wise, it'd be good to understand if all fields should be visible (even if disabled) in bulk editing vs only display those that can be edited as in this PR. cc @jameskoster 
- What makes a field not be editable in bulk? Specifically for this case: why we don't support bulk editing featured image, parent, or template? It seems to me that we shouldn't allow editing fields that are _unique_, like slug is — and they should be editable otherwise. This ties to validation as well. Or are there other cases in which we don't _want_ to allow fields editable in bulk?
- status doesn't work:

https://github.com/user-attachments/assets/0b638d59-e61a-4924-b21d-7ab259793e90

oandregal commented on packages/dataviews/src/normalize-fields.ts:
```diff
@@ -64,6 +64,11 @@ export function normalizeFields< Item >(
 			);
 		};
 
+		const supportsBulkEditing =
```
Why bulk editing is tied to the field type or even having a custom edit function? In my view, this is business/consumer logic: users can't edit fields that are unique. That is unrelated to any field type or what kind of edit function is in place.

I was looking for something like this in the fields package instead (in the slug field, for example). Additionally, if we agree that the only case in which a field cannot be bulk edited is when it needs to be unique, I wonder how can we tie this to validation. For example, when editing the field, the user shouldn't be able to add a value that is in use by another item.

oandregal replied:
The way I think about this is the following: any field is bulk editable regardless of their fieldType/edit function, unless it provides a specific opt-out (it's an unique field).

youknowriad replied:
Some fields could be unique for instance or things like that.

louwie17 committed:
Use DataFormControlPropsWithBulkEditing type instead to make type checking more robust

louwie17 committed:
Fix status field when switching to future

louwie17 commented:
> Sharing some early thoughts after having given this a quick test (haven't looked at the code yet):

Thanks :) I will be making some updates after Riad's review.

> * What makes a field not be editable in bulk? Specifically for this case: why we don't support bulk editing featured image, parent, or template? It seems to me that we shouldn't allow editing fields that are _unique_, like slug is — and they should be editable otherwise. This ties to validation as well. Or are there other cases in which we don't _want_ to allow fields editable in bulk?

Yeah good question, I think limiting to unique fields only makes sense and defaulting to fields supporting bulk editing. In terms of additional use cases, it may also be a question for @jameskoster. For example there is nothing limiting us from displaying the featured image field compared to the rest, but is this something users will actually use within bulk edit mode?

This also brings up the question of whether we want to disable bulk editing purely within the layout ( although the field may support it ). Something @gigitux suggested in my [last version](https://github.com/WordPress/gutenberg/pull/67344#issuecomment-2506418432) to move the `supportsBulkEditing` to the `SimpleFormField` type: https://github.com/WordPress/gutenberg/blob/0f3c00c4d51823246bce76e484ded97662934540/packages/dataviews/src/types.ts#L526-L530

> * status doesn't work:

Looks like I forgot to update this line with `editedRecord` as well: https://github.com/WordPress/gutenberg/pull/67584/files#diff-4ed486e973856f952d2b80aea381fc1f5a1d0004995a0b68554147b02cee0b55R106

louwie17 committed:
Default bulk editing support to true

jameskoster commented:
> UI-wise, it'd be good to understand if all fields should be visible (even if disabled) in bulk editing vs only display those that can be edited as in this PR. cc @jameskoster

I see this panel as an Inspector/Edit interface, so it seems reasonably clear to me that we should show uneditable fields while inspecting a single record. I don't have a strong feeling about bulk editing though, it may be something we decide field by field. For instance 'file size' in the media library should show a range based on the selection, which might be useful, whereas 'File type: Mixed' is less useful.

That said, it seems okay to hide uneditable fields during bulk selection for now :) 

> Yeah good question, I think limiting to unique fields only makes sense [...] For example there is nothing limiting us from displaying the featured image field compared to the rest, but is this something users will actually use within bulk edit mode?

Prohibiting bulk edits to unique fields seems like a good place to start. 

I agree it's unlikely a user would want to use the same featured image, title, etc. on multiple records, but I also acknowledge it's not impossible, so I don't mind allowing that initially.

louwie17 committed:
Fix conditionals within template field

louwie17 committed:
All template usage for bulk editing

gigitux commented:
> I think I'd start by implementing the `unique` prop: that's something we want and ties to other work we're already doing/planning to start around validation. Starting by `bulkEditing` would be ok, but I fear we'll be missing the bigger picture and the connection with validation.


This is a great way to connect the dots! Thanks for proposing this! I'm entirely agree with this approach!

louwie17 committed:
Move overlapping logic to useFieldValue hook

louwie17 committed:
Change naming of bulkEditing to unique

louwie17 commented on packages/edit-site/src/components/post-edit/index.js:
```diff
@@ -50,11 +68,20 @@ function PostEditForm( { postType, postId } ) {
+									id
+								)
+						  )
+						: null,
```
I will replace this with `getEditedEntityRecords` once https://github.com/WordPress/gutenberg/pull/67872 is merged.

louwie17 committed:
Update the loading state to account for bulk edits

louwie17 commented:
> I think I'd start by implementing the `unique` prop: that's something we want and ties to other work we're already doing/planning to start around validation. Starting by `bulkEditing` would be ok, but I fear we'll be missing the bigger picture and the connection with validation.

Yeah I am totally onboard with using the `unique` prop, as this does depict our current use case better.
I updated it as part of this commit: https://github.com/WordPress/gutenberg/pull/67584/commits/6c629e6a926ad9f2d8eba0c2f3564550850cf945, although I did have one follow up question @oandregal: When you are referring to the `Fields API`, do you mean the place where I have added now, as part of the `Field` type: https://github.com/WordPress/gutenberg/pull/67584/files#diff-05cd2939bbf237a5dbcd116a0da9c9019b05ec1ceaa9d429c5fc872f149f63c4R164-R168
Or adding it to the `SimpleFormField` type, which also defines the layout and label position? https://github.com/WordPress/gutenberg/blob/b9ca4814a364228bb786921186636b9b31a28aea/packages/dataviews/src/types.ts#L536-L540
Both places is also a valid answer :)

oandregal commented:
> When you are referring to the Fields API, do you mean the place where I have added now, as part of the Field type

Yes, what you did.

louwie17 committed:
Set postId to undefined if editing multiple records

louwie17 commented:
@gigitux, @youknowriad, and @oandregal I moved this PR out of draft now and should be ready for a re-review. There is just one thing that is dependant on the hook introduced in this PR: https://github.com/WordPress/gutenberg/pull/67872, that I will have to update once that is merged.

oandregal commented:
Took a look at this one. Haven't been able to review all use cases (combined fields).

| Function    | Input                                                             | Comment                                                                                                                                                                                                                                                                                                                                                                         |
| --- | --- | --- |
| `Edit`      | `data` (single item or array) + consolidated value (can be mixed) | |
| `render`    | `data` (single item)  | If there's multiple items, we display `MIXED` and do not call render. |
| `isVisible` | `data` (single item or array)   | Need to [fix type](https://github.com/WordPress/gutenberg/blob/022ef5973100c7f4bd58cd44eaaf677965eab812/packages/dataviews/src/types.ts#L132) to support arrays. Need to update all consumers that use it to suport arrays ([example](https://github.com/WordPress/gutenberg/blob/022ef5973100c7f4bd58cd44eaaf677965eab812/packages/fields/src/fields/password/index.tsx#L18)) or call `isVisible` for every item. |
| `isValid`   | `data` (single item) | This is called by the consumer via `isItemValid` utility ([code](https://github.com/WordPress/gutenberg/blob/022ef5973100c7f4bd58cd44eaaf677965eab812/packages/dataviews/src/validation.ts#L25)). What do we do with this one?  |

I wonder how can we improve the experience for consumers in the `Edit` function. Do we really need to pass both the `data` (single + array) and the consolidated `value`? We may run into the following situations:

1. DataForm with a single item. Edit receives a single item.
2. DataForm with multiple items and, for a given field, all are equal. Edit receives a single item.
3. DataForm with multiple items and, for a given field, there are mixed results. This is the crux we need to solve.

Thinking out loud about 3. We don't really know anything about the data `Edit` expects and how it'll operate with it. That's why mixed is difficult: we can't simple pass empty or null because those may have a special meaning for Edit. And that's the only case in which Edit would need access to the array of data.

**What if we just pass the array in the 3 scenario and we never pass `MIXED` to consumers?** If Edit receives an array, it needs to resolve the values itself. We'd also have to provide implementations for the controls we bundle (integer, datetime, etc.): we could set them to "empty value" in that scenario, for example.

louwie17 commented:
> Took a look at this one. Haven't been able to review all use cases (combined fields).

Thanks appreciate it :)

> `render` - `data` (single item) -  If there's multiple items, we display `MIXED` and do not call render.

Do note that for the `render` function the prop name is actually called `item` -> https://github.com/WordPress/gutenberg/blob/2e3e6e40fc8dac90a34bf880fe5749c638e4861d/packages/dataviews/src/types.ts#L190 which insinuates a single item. 

> Thinking out loud about 3. We don't really know anything about the data `Edit` expects and how it'll operate with it. That's why mixed is difficult: we can't simple pass empty or null because those may have a special meaning for Edit. And that's the only case in which Edit would need access to the array of data.

Yeah I agree, this is why I end up adding the `value` prop, which can be the a `MIXED` value when using TS. Of-course this would also require an explicit mention in the docs for consumers that don't use TS. If we stick with this.

> **What if we just pass the array in the 3 scenario and we never pass `MIXED` to consumers?** If Edit receives an array, it needs to resolve the values itself. We'd also have to provide implementations for the controls we bundle (integer, datetime, etc.): we could set them to "empty value" in that scenario, for example.

This approach would help simplify our logic and keep things consistent. It also forces 3rd party Edit functions to support "bulk editing" by default.
So in a way I did be ok going this route. The only thing that makes me feel hesitant is that for a form, handling an array of items seems like an edge case. For example we won't need this when using it in settings or the pattern creation form for example.

Or are you thinking that we did still allow users to pass in an object to DataForm -> `data: Item | Item[];` in `DataFormProps`, but that the `Edit`, `isVisible`, and `isValid` `data` props always accept an array of items? I would lean towards this approach.

oandregal commented:
I'd like to do the minimal thing that would work so we can merge and iterate. These are my current thoughts:

- `isValid`: stays as it is, it's up to the consumer to decide what to do.
- `isVisible`: it still receives a single item. But if DataForm receives an array, we run the visibility check for every item (all items in have to pass for it to be visible).

As for `Edit`, we aim to support bulk editing, so consumers would need to be aware of this. If field types work, most of the consumers wouldn't implement an Edit. The question is what's the API for it:

1. Pass item or array as data + the value that can be mixed (as currently implemented in this PR)
2. Pass item or array as data

For this PR, I'd simplify by going with the second option.

youknowriad commented:
> Pass item or array as data + the value that can be mixed (as currently implemented in this PR)
> Pass item or array as data

I would argue that for most situations, passing the "value" is the best thing to do. If you're a "string control" or a "number control" or whatever, you don't really care about the item or array of items, you just want to know the current value, or whether the value is mixed.

oandregal commented:
> I would argue that for most situations, passing the "value" is the best thing to do. If you're a "string control" or a "number control" or whatever, you don't really care about the item or array of items, you just want to know the current value, or whether the value is mixed.

I feel both passing only the `value` or only the `data` (item/array) can work. The later allows consumers to decide how to visualize data in a "mixed" scenario while the former doesn't (because they only have access to "mixed"). I'd be fine with any approach, though I'd prefer the later as more future-proof.

In terms of complexity for consumers: my thinking is that most of the use cases should be covered with our controls, so they won't deal with this. And when consumers write its own custom Edit function, they should be able to have the full power (access to data, not just an internal representation of "mixed").

louwie17 commented:
> I would argue that for most situations, passing the "value" is the best thing to do. If you're a "string control" or a "number control" or whatever, you don't really care about the item or array of items, you just want to know the current value, or whether the value is mixed.

I would agree with this as well and also agree with @oandregal that we did want to do "either or", as otherwise it is a bit confusing on whether we recommend using `value` or `data`.
If we do go with the value approach, it would be great to still provide some way for consumers to retrieve other values within the `Edit` function. Either by exposing `data` through the `DataFormContext` similar to how we are exposing it in the `DataViewsContext`, or providing a function/hook ( `useFieldValue` or `getFieldValue` that also handles the mixed ).

ntsekouras commented:
Thanks for the PR! Is this approach still relevant and we should refresh the PR or close it? It's been  a while since it has been updated.

Same question for the alternative lined PR: https://github.com/WordPress/gutenberg/pull/67344.
